netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -837,6 +827,105 @@ public void onPingTimeout() {
     }
   }
 
+  private final class GracefulShutdownRunner {
+
+    ChannelHandlerContext ctx;
+    long payload;
+    String goAwayMessage;
+
+    /**
+     * The grace time between starting graceful shutdown and closing the netty channel,
+     * {@code null} is unspecified.
+     */
+    @CheckForNull
+    Long graceTimeInNanos;
+
+    /**
+     * True if ping is Acked or ping is timeout.
+     */
+    boolean pingAckedOrTimeout;
+
+    /**
+     * Deadline of the shutdown.
+     */
+    long deadline;
+    Future<?> pingFuture;
+
+    GracefulShutdownRunner(
+        ChannelHandlerContext ctx, long payload, String goAwayMessage,
+        @Nullable Long graceTimeInNanos) {
+      this.ctx = ctx;
+      this.payload = payload;
+      this.goAwayMessage = goAwayMessage;
+      this.graceTimeInNanos = graceTimeInNanos;
+    }
+
+    /**
+     * Sends out first GOAWAY and ping, and schedules second GOAWAY and close.
+     */
+    void run() {
+      goAway(
+          ctx,
+          Integer.MAX_VALUE,
+          Http2Error.NO_ERROR.code(),
+          ByteBufUtil.writeAscii(ctx.alloc(), goAwayMessage),
+          ctx.newPromise());
+      ctx.flush();
+
+      long gracefulShutdownPingTimeout = GRACEFUL_SHUTDOWN_PING_TIMEOUT_NANOS;
+      if (graceTimeInNanos != null) {
+        deadline = ticker.read() + graceTimeInNanos;
+        gracefulShutdownPingTimeout = Math.min(gracefulShutdownPingTimeout, graceTimeInNanos);
+      }
+      pingFuture = ctx.executor().schedule(
+          new Runnable() {
+            @Override
+            public void run() {
+              secondGoAwayAndClose();
+            }
+          },
+          gracefulShutdownPingTimeout,
+          TimeUnit.NANOSECONDS);
+
+      encoder().writePing(ctx, false /* isAck */, payload, ctx.newPromise());
+      ctx.flush();
+    }
+
+    void secondGoAwayAndClose() {
+      if (pingAckedOrTimeout) {
+        return;
+      }
+      pingAckedOrTimeout = true;
+
+      checkNotNull(pingFuture, "pingFuture");
+      pingFuture.cancel(false);
+
+      // send the second GOAWAY with last stream id
+      goAway(
+          ctx,
+          connection().remote().lastStreamCreated(),
+          Http2Error.NO_ERROR.code(),
+          ByteBufUtil.writeAscii(ctx.alloc(), goAwayMessage),
+          ctx.newPromise());
+      ctx.flush();
```

Comment:
Ditto remove. You only need the flush in the gracefulShutdownPingTimeout handling (not for the ping receiving).